src/main/java/yanagishima/controller/ShareController.java

```diff
@@ -93,4 +126,55 @@ public void download(@RequestParam(name = "publish_id", required = false) String
     }
     return body;
   }
+
+  @PostMapping("share/updateViewers")
+  public Map<String, Object> updateViewers(@RequestParam(name = "publish_id") String publishId,
+                                           @RequestParam String viewers,
+                                           User user,
+                                           HttpServletResponse response) {
+    Map<String, Object> body = new HashMap<>();
+    try {
+      publishService.get(publishId).ifPresent(publish -> {
+        String publishUser = publish.getUser();
+        String requestUser = user.getId();
+        if (publishUser != null && publishUser.equals(requestUser)) {
+          publishService.update(publish, viewers);
+          body.put("viewers", viewers);
+          emitPublishEvent(publish.getPublishId(), publish.getDatasource(), publish.getEngine(),
+                  publish.getQueryId(), publish.getUser(), viewers);
+        } else {
+          AccessControlUtil.sendForbiddenError(response);
+          return;
+        }
+      });
+    } catch (Throwable e) {
+      log.error(e.getMessage(), e);
+      body.put("error", e.getMessage());
+    }
+    return body;
+  }
+
+  private void emitPublishEvent(String publishId, String datasource, String engine, String queryId, String user,
+                                String viewers) {
+    Map<String, Object> event = new HashMap<>();
+    event.put("publish_id", publishId);
+    event.put("datasource", datasource);
+    event.put("engine", engine);
+    event.put("query_id", queryId);
+    event.put("user", user);
+    event.put("viewers", viewers);
+    fluencyClient.emitPublish(event);
+  }
+
+  private boolean canAccessPublishedPage(String publishUser, String requestUser, String viewers) {
+    if (publishUser != null && publishUser.equals(requestUser)) {
+      return true;
+    }
+
+    if (viewers != null && viewers.contains(requestUser)) {
```

Comment:
Yeah, I guess this logic is a little controversial.
I expect format of viewers param is like p12345,jp67890 even though there is no specific pattern in server-side.

client side prepares textarea and user inputs.
client side checks user inputs per line and sends commna separated viewer list.

Let's suppose user jp11111 publishes and then add jp12345 and jp67890 as viewers.
In this case, user jp11111, jp12345, jp67890 can access published page but user jp99999 can't access because jp99999 is neither publisher nor viewer.

refactor and add test in https://github.com/yanagishima/yanagishima/pull/305/commits/a8062c0907c41855d1be9e995c06a3258bf67687

Reply:
Sorry for my lack of explanation. I meant short user id may hit accidentally. 

User  `jp123` can access the page when viewers field is `jp12345,jp67890`. If frontend requests with `,` separator, it would be better to split in server side. 

Another question is case sensitivity. Will frontend lowercase `viewers` in their code?

Reply:
ah, you are right, great catch!
fixed in https://github.com/yanagishima/yanagishima/pull/305/commits/d9d01abbc31134c9a084fdfe4823038719c19003

> Another question is case sensitivity. Will frontend lowercase viewers in their code?

Yes, but just in case, set lower in https://github.com/yanagishima/yanagishima/pull/305/commits/c90523c51d75de73efbdbff77c4add4d1a70152a